[Tooling] Add Binary Compatibility Validator & Review Public Contract

ParaskP7 opened the pull request:
Closes: AINFRA-868 (h/t @wzieba)

---

### Description

This PR adds the [binary compatibility validator](https://github.com/Kotlin/binary-compatibility-validator) to help reduce risks of introducing unwanted public API changes, as well as review the latest state of the public contract of this library, then update it as needed before the `1.0.0` release (see p1751457120601209/1748330120.391159-slack-C030U03RC8Y).

---

### Testing Steps

Using the [DOAndroid](https://github.com/bloom/DayOne-Android) or/and [PCAndroid](https://github.com/Automattic/pocket-casts-android) project, update `automattic-encryptedlogging` to `20-ef2bdcf0cca59bf0cc0b15d48866666d6f48b0fa` (version which is based on this PR's [latest](https://github.com/Automattic/EncryptedLogging/pull/20/commits/ef2bdcf0cca59bf0cc0b15d48866666d6f48b0fa) commit) and make sure that encrypted logging continues to work as expected:

1. Using `App Inspection` from AS find the `encrypted-logging.db` and its sole `EncryptedLogModel` table. Notice that the table is empty, there are no log to be sent.
2. Make the app crash. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could navigate to `Settings` -> `Developer` and then click `Crash The App`.
3. You might now notice a new db entry within `App Inspection` from AS on that `DETACHED` app (or not, it depends). Now, switch to the `ATTACHED` app and notice the db empty. This is because this db entry/log has been already sent to Sentry already, on app start and after the app crashed. To really notice the entry, trying crashing the app again, this time put the device on offline mode first. Doing that you'll be able to see the entry on the db, even when switching to the newly created `ATTACHED` app.
4. Open [Sentry](https://a8c.sentry.io) and navigate to `Issues`. Find the project you're testing for (for example `day-one-android` or/and `pocket-casts-android`) and check the latest `debug` exceptions. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could search for `Developer caused a crash!` and check the latest crash. Now navigate at the bottom and check the `uuid`. Verify it is the same `UUID` that you saw on that db entry before.
5. Open [Read Encrypted Logs](https://mc.a8c.com/encrypted-logs.php), copy-paste that `UUID` into `Log UUID`, click `View` and make sure you can read the newly uploaded and now decrypted log.

---

Additionally, run this [ReleaseStack_EncryptedLogTest](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/androidTest/kotlin/com/automattic/encryptedlogging/release/ReleaseStack_EncryptedLogTest.kt#L43)  instrumented test locally, and make sure that it runs successfully, and as expected. Note that:

1. It is quite hard to work with this test because `TooManyRequests` can be hit quite easily. At which point you need to wait for a few minutes (15' minutes to half an hour) before being able to verify it is working as expected.
2. In order to verify that this test works as expected one would need to use the debugger and navigate through the flow manually. Otherwise the test might success with a false positive, but actually silently failing on (at least) this `TooManyRequests` problem mentioned above.

ParaskP7 committed:
Build: Add Binary Compatibility Validator & Review Public Contract

This 'encryptedlogging.api' file got auto-generated: ./gradlew apiDump

ParaskP7 committed:
Refactor: Remove double exclamation marks from everywhere

ParaskP7 committed:
Api: Replace unnecessary public apis with internal

Then re-generate 'encryptedlogging.api' using: ./gradlew apiDump

ParaskP7 committed:
Test: Replace let with check not null on encrypted log dao test

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/22
#discussion_r2180028059

ParaskP7 commented on encryptedlogging/src/test/kotlin/com/automattic/encryptedlogging/persistence/dao/EncryptedLogDaoTest.kt:
```diff
@@ -69,7 +69,7 @@ class EncryptedLogDaoTest {
                 .ignoringFields("id")
                 .isEqualTo(logsToBeInserted[index])
-            sut.deleteEncryptedLog(log!!) // So that the next encrypted log can be fetched correctly
+            log?.let { sut.deleteEncryptedLog(it) } // So that the next encrypted log can be fetched correctly
```
Thanks 🤖 and done: b4b1737ff81bf3fd17905012e2489f69d29e5933

wzieba replied:
Hm, my concern with the suggested solution is that we would crash the app if `log` will be `null`. I think that encrypted logging isn't that crucial - could we change it to `if null: log a warning/error; else: delete the log`? WDYT?

ParaskP7 replied:
What do you mean by that @wzieba, all other tests still run fine? 🤔

![image](https://github.com/user-attachments/assets/56554835-589d-4987-b2c7-fc3b6143f87a)

![image](https://github.com/user-attachments/assets/d2fb0fea-e38c-44c9-8a8d-3df83b4e35e9)

wzieba replied:
ah sorry @ParaskP7 🤦 this is a test

ParaskP7 replied:
Yea, sorry for confusing you on that commit that touched both main/test code... 🫣

wzieba replied:
my bad, thanks for clarification 🙌

wzieba commented on encryptedlogging/api/encryptedlogging.api:
```diff
@@ -0,0 +1,22 @@
+	public fun uploadEncryptedLogs ()V
+}
+
+public final class com/automattic/encryptedlogging/BuildConfig {
+	public static final field APP_SECRET Ljava/lang/String;
+	public static final field BUILD_TYPE Ljava/lang/String;
+	public static final field DEBUG Z
+	public static final field ENCRYPTION_KEY Ljava/lang/String;
+	public static final field LIBRARY_PACKAGE_NAME Ljava/lang/String;
+	public fun <init> ()V
+}
```
It's great to have this `*.api` file! The `BuildConfig` is not needed - we only use it for tests. Could we please enable `BuildConfig` only for tests (maybe another variant? or only in debug?) and disable it in release?

ParaskP7 replied:
> It's great to have this *.api file!

Indeed, thanks again @wzieba ! 💪

> The BuildConfig is not needed - we only use it for tests.

Yea, that's why I left it there... 😊

> Could we please enable BuildConfig only for tests (maybe another variant? or only in debug?) and disable it in release?

Yea, I was thinking about that, wasn't sure how to go about it, let me see if your suggestion would work... 🤞

ParaskP7 replied:
So, coming back to you on that @wzieba :

1. This `android.buildFeatures.buildConfig` is module specific and thus cannot be enabled only for tests or a specific build variant.
2. We could move those `buildConfigField` into the `buildTypes.debug` configuration, which I think we should do anyway, but this will still cause `./gradlew apiCheck` to fail if `BuildConfig` is not included in that `*.api` configuration.
3. The only possible solution here is to move this whole `androidTest` configuration on another module, but I think that this is an overkill, at least for now.

Your thoughts? 🤔

wzieba replied:
Let's disable the task then, this works fine 👍 

```
tasks.whenTaskAdded {
    if (name == "generateReleaseBuildConfig") {
        enabled = false
        println("✅ Disabled task: $name")
    }
}
```

ParaskP7 replied:
I am not sure I understand what you are trying to do here @wzieba, can help clarify? 🙏

wzieba replied:
The above configuration will remove `BuildConfig` from being shipped with the release artifact. `BuildConfig` is not something we should offer for SDK consumers, it's not useful for them and might be confusing.

ParaskP7 replied:
I still don't get it @wzieba, apologies, `./gradlew apiCheck` would still fail, wouldn't it? 🤔

Maybe if you could sent me a patch of the complete solution you have in mind... 🙏

wzieba replied:
I've prepared a PR to validate it works on CI https://github.com/Automattic/EncryptedLogging/pull/23 , WDYT?

ParaskP7 replied:
Aw, interesting, it works! 😅 I could swear I tried that and it just didn't work, whooot... 😊 🫣 Thanks @wzieba! 💪

ParaskP7 replied:
We can now merge this, and then your #23, wdyt?

wzieba replied:
Yup! Maybe it wasn't working due to previous local state (should work after `./gradlew clean`)

ParaskP7 replied:
Yea, maybe, or me trying various things and making a mess... 🤷 No matter, success! 🥇

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/model/encryptedlogging/EncryptedLog.kt:
```diff
@@ -19,9 +19,8 @@ internal data class EncryptedLog(
-                // Crash if values are missing which shouldn't happen if there are no logic errors
-                uuid = encryptedLogEntity.uuid!!,
-                file = File(encryptedLogEntity.filePath),
+                uuid = checkNotNull(encryptedLogEntity.uuid) { "UUID cannot be null" },
+                file = File(checkNotNull(encryptedLogEntity.filePath) { "File path cannot be null" }),
```
If `uuid` cannot be null, can we update the database schema? The same goes for `file`

ParaskP7 replied:
Yea @wzieba , this got reviewed so many times and I didn't want to change that, just because of this comment on top of that, plus thinking you are aware of that and know this to be an edge case: 🤷

`// Crash if values are missing which shouldn't happen if there are no logic errors`

Do you want me to update `EncryptedLogEntity` as part of this and remove the defaults from that data class, just like we've start doing with the rest of such WellSQL migrations on WCAndroid? 🤔

wzieba replied:
yeah, I think if we crash the app if the `uuid` is `null`, then we shouldn't allow for this state to be possible 👍

ParaskP7 replied:
Ok, I'll proceed with this change too... 🙃

ParaskP7 replied:
FYI: This is now done: 0898a69924ae1988eef55c92c0cb3bb311c1e98d Cc @wzieba

ParaskP7 committed:
Migration: Make all encrypted log entity fields non-nullable

Plus, this change removes all default values per field, those are not
actually needed and make this entity instance creation unnecessarily
implicit.

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/22
#discussion_r2180098044

wzieba commented on encryptedlogging/api/encryptedlogging.api:
```diff
@@ -0,0 +1,22 @@
+	public fun enqueueSendingEncryptedLogs (Ljava/lang/String;Ljava/io/File;Z)V
+	public fun resetUploadStates ()V
+	public fun uploadEncryptedLogs ()V
+}
```
We could also make `AutomatticEncryptedLogging` internal: there's no value for consumers to access this class. I mean something like this b6ceaa7a6b1025bb8a7d2e2faef829a9b013b4b1 <- this will also make `AutomatticEncryptedLogging` a singleton, what is important as during its creation a new request queue is created e.g.

Hiding internal implementation classes is also in sync with other SDKs:
- `experimentation` hides `ExPlat` class  https://github.com/Automattic/Automattic-Tracks-Android/blob/30b9bb07d48ebfbcb68b33f39c46c56a6978431b/experimentation/src/main/java/com/automattic/android/experimentation/VariationsRepository.kt#L52
- `parsely` hides `ParselyTrackerInternal` class https://github.com/Parsely/parsely-android/blob/main/parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.kt#L162

ParaskP7 replied:
Oh, interesting, I like it @wzieba, let me apply this change too... 😅

ParaskP7 replied:
FYI: This is now done: fb44960f35150270cc972fa72d4b4853779dacfd Cc @wzieba

ParaskP7 committed:
Refactor: Make automattic encrypted logging an internal singleton (ish)

PS Discussion: https://github.com/Automattic/EncryptedLogging/pull/22
#discussion_r2180300989
